test/Microsoft.AspNetCore.Session.Tests/SessionTests.cs

```diff
@@ -667,7 +667,25 @@ public class SessionTests
                 response.EnsureSuccessStatusCode();
             }
 
-            var sessionLogMessages = sink.Writes.OnlyMessagesFromSource<SessionMiddleware>().ToArray();
+            var retries = 0;
+            WriteContext[] sessionLogMessages = null;
+            while(true)
+            {
+                try
+                {
+                    sessionLogMessages = sink.Writes.OnlyMessagesFromSource<SessionMiddleware>().ToArray();
+                    break;
+                }
+                catch (InvalidOperationException)
```

Comment:
This issue doesn't have anything to do with session does it, it's not the session logs that are causing this race, it's the hosting logs. The session logs are always written before the pipeline unwinds and the (empty) response returns (https://github.com/aspnet/Session/blob/dev/src/Microsoft.AspNetCore.Session/SessionMiddleware.cs#L120). It's Hosting/TestHost logging after the response is finished (https://github.com/aspnet/Hosting/blob/dev/src/Microsoft.AspNetCore.Hosting/Internal/HostingApplication.cs#L65).

Maybe we need to fix this in TestSink instead. If you add a lock in TestSink.Write and move OnlyMessagesFromSource directly onto TestSink and then add a lock in it, that should prevent this for everyone. OnlyMessagesFromSource may also need to incorporate the ToArray inside the lock.
https://github.com/aspnet/Logging/blob/9506ccc3f3491488fe88010ef8b9eb64594abf95/src/Microsoft.Extensions.Logging.Testing/TestSink.cs#L9